keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,405 @@
+prevents the usage of `7g.40gb` paritions and puts a tighter cap on `4g.20gb`
+partitions than would be implied by the `accelerator-memory` quota:
+```yaml
+...
+  - coveredResources: ["cpu", "memory", "kueue.x-k8s.io/accelerator-memory"]
+    flavors:
+    - name: default-flavor
+      resources:
+      - name: "cpu"
+        nominalQuota: ...
+      - name: "memory"
+        nominalQuota: ...
+      - name: kueue.x-k8s.io/accelerator-memory
+        nominalQuota: 200G
+      - name: nvidia.com/mig-1g.5gb
+        nominalQuota: 40  # 200G/5g
+      - name: nvidia.com/mig-2g.10gb
+        nominalQuota: 20  # 200G/10g
+      - name: nvidia.com/mig-4g.20gb
+        nominalQuota: 4  # less than the 10 that would be allowed by the acelerator-memory quota
+      - name: nvidia.com/mig-7g.40gb
+        nominalQuota: 0  # not allowed -- inefficient use of GPU resources
+```
+
+[mig]: https://docs.nvidia.com/datacenter/cloud-native/kubernetes/latest/index.html#the-mixed-strategy
+[instaslice]: https://github.com/openshift/instaslice-operator
+
+#### Story 2
+
+This KEP would allow Kueue to augment a Job's resource requests
+for quota purposes with additional resources that are not actually 
+available in the Nodes of the cluster. For example, it could be used to 
+define quotas in terms of a fictional currency `kueue.x-k8s.io/credits`
+to give "costs" to various resources. This could be used to approximate
+the actual monetary cost of a cloud resource (for example a larger or more powerful GPU
+could cost twice as many `credits` as a smaller or previous generation GPU).
+We'd expect in this usage pattern that `Replace` would be used so that
+resource list is just augmented with the computed `credits` resource.  We'd also
+expect that a ClusterQueue's `borrowingLimit` for `credits` would be set to `0`
+to enforce the intended fiscal constraint.
+
+## Design Details
+
+The only non-trivial design question is how best to specify the configuration.
+
+### Option 1: Singleton Matrix Input Focused
+The entire transformation matrix is defined in one object.  Most likely
+this object would be embedded inside Kueue's `Configuration` API, but
+it would also be possible to make it a separate top-level config API or 
+use a config map containing a single yaml field that would be deserialized
+into the type defined below.
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  ConversionBaseValues map[corev1.ResourceName ]resource.Quantity `json:"conversionBaseValues"`
+
+  // should the input resource be replaced or retained
+  Strategy ResourceTransformationStrategy `json:"strategy"`
+}
+
+type ResourceTransformationMatrix struct {
```

Comment:
Questions:
- Have you considered exposing the results of the transformations at the API level (for debuggability)?
- Will the used output resources be accounted for in the statuses of Workload (admission) and ClusterQueue?

Reply:
Exposing the details of the transformation for debugging is an interesting idea.  I had not thought of that. One possibility would be that if workload's resources are actually transformed then we could emit an event that contained the workload, and the input and output resources.  

I think for the second question the answer is that the output resources are treated no differently than any other (non-transformed) resource that is present in the Workload's Spec.PodSet[].  The rest of Kueue is not aware of whether a resource is in a PodSet because it was part of the "raw" PodSet that came from the result of GenericJob.PodSet() when the Workload was being constructed or because it was added by AdjustResources.

Reply:
> Exposing the details of the transformation for debugging is an interesting idea.

The more info we can expose without compromising the performance the better. If some requires significant work we could add points the the graduation criteria for second phase of Beta, or GA.

> if workload's resources are actually transformed then we could emit an event that contained the workload, and the input and output resources. 

That is a possibility, feel free to add, but I was thinking about exposing the information about the "effective" quotas after transformation. For example, from the story 1:

```yaml
  - coveredResources: ["cpu", "memory", "kueue.x-k8s.io/accelerator-memory"]
    flavors:
    - name: default-flavor
      resources:
      - name: "cpu"
        nominalQuota: ...
      - name: "memory"
        nominalQuota: ...
      - name: kueue.x-k8s.io/accelerator-memory
        nominalQuota: 320G
```
is transformed into 
```yaml
  - coveredResources: ["cpu", "memory", "kueue.x-k8s.io/accelerator-memory"]
    flavors:
    - name: default-flavor
      resources:
      - name: "cpu"
        nominalQuota: ...
      - name: "memory"
        nominalQuota: ...
      - name: kueue.x-k8s.io/accelerator-memory
        nominalQuota: 200G
      - name: nvidia.com/mig-1g.5gb
        nominalQuota: 40  # 200G/5g
      - name: nvidia.com/mig-2g.10gb
        nominalQuota: 20  # 200G/10g
      - name: nvidia.com/mig-4g.20gb
        nominalQuota: 10  # 200G/20g
      - name: nvidia.com/mig-7g.40gb
        nominalQuota: 5  # 200G/40g
```
But the lower yaml only is represented in memory, IIUC? So with this KEP the "spec" quotas are not actually the "effective" quotas. I would like to consider exposing the "effective" quotas. ClusterQueue status seems like a natural place. The transformations are not just additive so we would need to expose the whole effective structure. I'm wondering if we should do this unconditionally then.

> I think for the second question the answer is that the output resources are treated no differently than any other (non-transformed) resource that is present in the Workload's Spec.PodSet[]. The rest of Kueue is not aware of whether a resource is in a PodSet because it was part of the "raw" PodSet that came from the result of GenericJob.PodSet() when the Workload was being constructed or because it was added by AdjustResources.

I see, so in case there is a workload which consumes "nvidia.com/mig-1g.5gb" it will also consume 5 of `kueue.x-k8s.io/accelerator-memory`. And, we will see both resources also as "used" in the ClusterQueue status under "FlavorUsage](https://github.com/kubernetes-sigs/kueue/blob/baba452433312a0195bd9972c95a1c097d8de269/apis/kueue/v1beta1/clusterqueue_types.go#L320-L330)? Is my understanding correct?

Reply:
Thanks for the detailed comment.   I can see where I need to clarify the KEP.  The transformations are only applied to the Workload, not to the ClusterQueue.

I'm going to give a more detailed response in the KEP by expanding the MIG user stories with a concrete example of what a Workload and a ClusterQueue with that workload admitted would like under two different transformation configurations.  Will tag you when when I push updated version.

Reply:
Pushed a major revision.  I'm going to be offline Friday-Sunday so apologies in advance for being slow in responding to any additional comments you might have.